In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
y2015 = pd.read_csv('https://www.dropbox.com/s/0so14yudedjmm5m/LoanStats3d.csv?dl=1', skipinitialspace=True, header=1)
print('Number of columns:', len(y2015.columns))

C:\Users\James\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2698: DtypeWarning: Columns (0,19,55) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Number of columns: 111


In [3]:
# Convert ID and Interest Rate to numeric.
y2015['id'] = pd.to_numeric(y2015['id'], errors='coerce')
y2015['int_rate'] = pd.to_numeric(y2015['int_rate'].str.strip('%'), errors='coerce')

# Drop other columns with many unique variables
y2015.drop(['url', 'emp_title', 'zip_code', 'earliest_cr_line', 'revol_util', 'sub_grade', 'addr_state', 'desc'], 1, inplace=True)

# Remove two summary rows at the end that don't actually contain data.
y2015 = y2015[:-2]

In [4]:
# Create a correlation matrix and delete closely correlated columns (> 0.5).
col_corr = set() 
corr_matrix = y2015.corr().abs()
for i in range(len(corr_matrix.columns)):
    for j in range(i):
        if corr_matrix.iloc[i, j] >= 0.5:
            colname = corr_matrix.columns[i] 
            col_corr.add(colname)
            if colname in y2015.columns:
                del y2015[colname] 
corr_matrix

id  member_id  loan_amnt  funded_amnt  \
id                              1.000000   0.997540   0.008288     0.008288   
member_id                       0.997540   1.000000   0.008620     0.008620   
loan_amnt                       0.008288   0.008620   1.000000     1.000000   
funded_amnt                     0.008288   0.008620   1.000000     1.000000   
funded_amnt_inv                 0.008554   0.008885   0.999994     0.999994   
int_rate                        0.053402   0.053447   0.140572     0.140572   
installment                     0.012919   0.013274   0.941205     0.941205   
annual_inc                      0.015055   0.014958   0.305734     0.305734   
dti                             0.005692   0.005806   0.006103     0.006103   
delinq_2yrs                     0.002300   0.002261   0.010424     0.010424   
inq_last_6mths                  0.000660   0.000715   0.039040     0.039040   
mths_since_last_delinq          0.006322   0.006186   0.034863     0.034863   
mths_since_last_record          0.037319   0.037366   0.014108     0.014108   
open_acc                        0.006309   0.006111   0.188214     0.188214   
pub_rec                         0.003501   0.003635   0.088899     0.088899   
revol_bal                       0.003774   0.003745   0.334656     0.334656   
total_acc                       0.014479   0.014673   0.208430     0.208430   
out_prncp                       0.192187   0.191883   0.623926     0.623926   
out_prncp_inv                   0.192058   0.191755   0.623961     0.623961   
total_pymnt                     0.219883   0.219945   0.706184     0.706184   
total_pymnt_inv                 0.220078   0.220139   0.706157     0.706157   
total_rec_prncp                 0.188353   0.188443   0.556266     0.556266   
total_rec_int                   0.165885   0.165861   0.736585     0.736585   
total_rec_late_fee              0.015779   0.015748   0.051888     0.051888   
recoveries                      0.070027   0.069799   0.077870     0.077870   
collection_recovery_fee         0.070594   0.070372   0.079190     0.079190   
last_pymnt_amnt                 0.068844   0.068974   0.259864     0.259864   
collections_12_mths_ex_med      0.011953   0.011891   0.027454     0.027454   
mths_since_last_major_derog     0.021221   0.021181   0.018556     0.018556   
policy_code                          NaN        NaN        NaN          NaN   
...                                  ...        ...        ...          ...   
mo_sin_old_rev_tl_op            0.004593   0.004731   0.183754     0.183754   
mo_sin_rcnt_rev_tl_op           0.009478   0.009476   0.066548     0.066548   
mo_sin_rcnt_tl                  0.005305   0.005469   0.041686     0.041686   
mort_acc                        0.013036   0.013261   0.231574     0.231574   
mths_since_recent_bc            0.004085   0.004029   0.051817     0.051817   
mths_since_recent_bc_dlq        0.005001   0.005281   0.057153     0.057153   
mths_since_recent_inq           0.009252   0.009397   0.008587     0.008587   
mths_since_recent_revol_delinq  0.001272   0.001500   0.049828     0.049828   
num_accts_ever_120_pd           0.006172   0.006180   0.057356     0.057356   
num_actv_bc_tl                  0.006015   0.005916   0.212990     0.212990   
num_actv_rev_tl                 0.014951   0.014976   0.164565     0.164565   
num_bc_sats                     0.008984   0.008889   0.222171     0.222171   
num_bc_tl                       0.028205   0.028132   0.203509     0.203509   
num_il_tl                       0.000665   0.000371   0.074857     0.074857   
num_op_rev_tl                   0.000275   0.000317   0.174867     0.174867   
num_rev_accts                   0.017973   0.017942   0.175080     0.175080   
num_rev_tl_bal_gt_0             0.013698   0.013587   0.166295     0.166295   
num_sats                        0.005433   0.005237   0.189328     0.189328   
num_tl_120dpd_2m                0.005254   0.005197   0.000905     0.000905   
num_tl_30

In [5]:
print('Number of columns:', len(y2015.columns))

Number of columns: 47


The number of columns dropped from the original 111 to 47 after dropping related features. The remaining columns are shown below.

In [6]:
for i in y2015.columns:
    print(i)

id
loan_amnt
term
int_rate
grade
emp_length
home_ownership
annual_inc
verification_status
issue_d
loan_status
pymnt_plan
purpose
title
dti
delinq_2yrs
inq_last_6mths
mths_since_last_record
open_acc
pub_rec
revol_bal
initial_list_status
total_rec_late_fee
recoveries
last_pymnt_d
next_pymnt_d
last_credit_pull_d
collections_12_mths_ex_med
policy_code
application_type
dti_joint
verification_status_joint
acc_now_delinq
tot_coll_amt
open_acc_6m
open_il_12m
mths_since_rcnt_il
il_util
inq_fi
total_cu_tl
chargeoff_within_12_mths
delinq_amnt
mo_sin_old_il_acct
mo_sin_old_rev_tl_op
mo_sin_rcnt_rev_tl_op
num_accts_ever_120_pd
num_tl_120dpd_2m


In [7]:
# Run the Random Forest classifier.
from sklearn import ensemble
from sklearn.model_selection import cross_val_score

rfc = ensemble.RandomForestClassifier()
X = y2015.drop('loan_status', 1)
Y = y2015['loan_status']
X = pd.get_dummies(X)
X = X.dropna(axis=1)

print('Scores:', cross_val_score(rfc, X, Y, cv=10))
print('\n')
print('Average score:', cross_val_score(rfc, X, Y, cv=10).mean())

Scores: [ 0.93087317  0.93244045  0.924794    0.9214457   0.89693659  0.92707195
  0.90180247  0.93863545  0.94483103  0.94786966]


Average score: 0.911351682954


The model still scored above 90% after dropping more than half of the features. The following columns, which are payment amount or outstanding principal related, were dropped during the correlation matrix run with no adverse impact to the model: out_prncp, out_prncp_inv, total_pymnt, total_pymnt_inv, installment, total_rec_prncp, and last_pymnt_amnt.